### ISSUE TITLE
Bug with gpx tracks where lon<0 and lat <0: adding a track not finishing

### ISSUE THREAD

**Issue by keenonkites** [Maproom/qmapshack#229]
**Important Notice:**

**Do not delete any of the sections. Answer them. If you really**
**think they do not apply to you give a short statement. Bug**
**reports that are not properly filed might be closed without**
**further notice.**

**Choose a good title**

What is a good title?
The title will be listed in the changelog when the issue is fixed. Do
you think someone else would understand what's fixed or done from your
title? Does it really name the real topic of the issue?

[yes ] Did you choose a good title?

**Describe the bug**
Running on Windows 10 64bit, qmapshack v1.15.0 64 bit.
I have gpx tracks (Garmin) from Argentina where, obviously, both lat and lon are below zero. Trying to add them to the workspace (drag and drop) does not end properly (screenshot). It shows a new import but without any names and nothing is usable of this track. Any other actions in the workspace (like doubleclick on a visible track) does not execute any actions.
When restarting qmapshack (this works normally, no Task Termination needed normally), the problematic track is not anymore on the workspace.

**What have you done to circle down the problem?**
I've taken an existing Track from Argentina and stripped it down (Text Editor) to 2 Trackpoints to play around. I've changed the coordinates by deleting the dash '-' in front of it, to make the values positive. The Problem exists only if both lat and lon are negative, any other combinations work as expected (only one negative value, both positive)

Did you compare to previous versions? What versions? Did you check
if others have the same problem?
No

**To Reproduce**

1. Download attached zip, extract tracks
2. Drag and Drop test-crash.gpx to QMapShack (I did drag it normally to the Map part of the app)
3. Import hangs, QMapShack not usable anymore, restart needed
4. using test-crash-fixed.gpx (where I just changed lon from negative to positive) for the same test works properly, no hang, no problem

**Expected behavior**
Finish Import and have a properly usable Track as with other Tracks

**Screenshots**
Half imported Track -> restart of app needed
![image](https://user-images.githubusercontent.com/3951936/88395658-b65aaf00-cdc1-11ea-8a34-fa2e55c30b15.png)

**Attachments**
[test-gpx-files.zip](https://github.com/Maproom/qmapshack/files/4972159/test-gpx-files.zip)

**Tracebacks**

If QMapshack crashes the fastest way to get help is a backtrace. We are
aware that it's not that easy to record one. However if you provide one
chances are good to have it solved very fast. If the problem can't be
reproduced on the developer's environment we will sooner or later request
a traceback.

For Linux see:
https://github.com/Maproom/qmapshack/wiki/TroubleShooting#create-a-backtrace-of-a-crash-on-linux

For Windows it would need a debug build and running QMapshack in Visual Studio.

**Desktop (please complete the following information):**

 - OS: Windows 10 64bit
 - QMapShack Version1.15.0 64bit

**Additional context**

**Comment by kiozen** [comment]
I can't reproduce the bug. I can load test-crash.gpx without a problem. I see a track with two waypoints somewhere in the south west.

On a restart the track is restored properly. 

There must be something else wrong. If this would be a general issue I am sure someone would already have complained in the last years

**Comment by keenonkites** [comment]
I've tested it with several newer and older Argentinian GPX Tracks, it always happens. The track uploaded is just a manually created/edited testtrack.

Is there a way to enable debugging on Windows 10 qmapshack for narrowing down this problem and see if it's a local problem on my installation ?

To be honest, I thought the same: if about a quarter of the world would not be able to use QMapShack you would have heard it before... ;-)

Thanks for your help.

**Comment by keenonkites** [comment]
When starting QMapShack with -f I see only a fewlines relevant.... all the rest is startup protocoll.

Drag&Drop my problematic file
...
2020-07-25 11:12:13.096 [debug] CGisItemTrk::updateVisuals() "CrashTest" "setupInterpolation()"
2020-07-25 11:12:13.096 [debug] CGisItemTrk::updateVisuals() "CrashTest" "deriveSecondaryData()"

If I then close the app I get following error messages, most probably only follow up errors due to original problem:
2020-07-25 11:15:23.267 [debug] slotSaveWorkspace()
2020-07-25 11:15:23.596 [warning] Execution of SQL-Statement ` "INSERT INTO workspace (type, keyqms, name, changed, visible, data) VALUES (:type, :keyqms, :name, :changed, :visible, :data)" ` failed:
2020-07-25 11:15:23.596 [warning] QSqlError("19", "Unable to fetch row", "NOT NULL constraint failed: workspace.name")
2020-07-25 11:15:23.693 [debug] CWptIconManager::~CWptIconManager()


When using the fixed version in the zip file I have three entries:
...
2020-07-25 11:16:37.853 [debug] CGisItemTrk::updateVisuals() "CrashTest" "setupInterpolation()"
2020-07-25 11:16:37.853 [debug] CGisItemTrk::updateVisuals() "CrashTest" "deriveSecondaryData()"
2020-07-25 11:16:37.855 [debug] CGisItemTrk::updateVisuals() "CrashTest" "findWaypointsCloseBy()"

When closing the app after adding the good file there are now errors in the logfile.

Probably this helps.... can the log level be increased ?

**Comment by keenonkites** [comment]
After further investigations I found out:

- I get also around the problem when deleting tiem <time>xxxx</time> Values inside the <trkpt ...> parts
- I get also around the problem when switching (inside QMapShack) the timezone setup to either 'Local' or 'UTC'
- I get the problem when switching the timezone setup to automatic (as it was initially in my app) or fix to 'America/Argentina/Buenos_Aires'
- I was not even able to open my workspace (with lots of local tracks added) when having timezone setup fixed to 'America/argentina/Buenos_Aires', had to create an empty one to be able to switch back
- Doing the same tests on MacOS Catalina does not show the same behaviour, I can always add the problematic track, whatever is set in timezone setup

Further Information: my local timezone is Europe/Zurich, I assume. I'm running on Swiss Timezone on my Windows.
![timezonesetup](https://user-images.githubusercontent.com/3951936/88458869-4582c800-ce91-11ea-988e-edf701ccd38f.jpg)
e. 

It looks like something (my machine only, specific constellation of Windows 10, QMapShack) has problems interpreting this specific timezon

Having the timezone set to 'Automatic' does, I guess at least' after seeing actual behaviour, calculate the needed timezone out of the lat and lon values..... this would explain why it suddenly works after changing one negative value into a positive one: that get's you for sure into a different timezone.

Would you mind testing that again on your Windows 10 (I assume) installation in order to confirm if it's something only on my side, inside Windows somewhere or nevertheless inside QMapShack.

I hope the additional Information help to nail down/fix the issue, wherever it stays.

Sorry for the hassles and many thanks for your support.

**Comment by wthaem** [comment]
Tested with
* Windows10, OS system timezone: Berlin/Germany
* QMS 1.15.0, QMS timezone set to `America/Argentina/Buenos_Aires`

Result: no problem with `test-crash.gpx`. GPX file is properly loaded. Can close and restart QMS - no problems.

**Comment by kiozen** [comment]
Same here. No problem. (Linux)

**Comment by keenonkites** [comment]
@wthaem 
the problem is not closing and restarting QMS, the problem is that the problematic track is not added properly, the process is not finished properly. But the application can be restarted without any problem, just the added 'problematic' track is gone again. But you state that the GPX file is properly loaded.

@kiozen 
I've installed QMS on a complete different Windows 10 64 and ran into the exact same problem:

1. Downloaded QMS Win64 1.15.0 installer
2. Executed the installer and selected Installation of MSVC11 2017 Runtime too
3. MSVC11 2017 runtime could not be installed as another version is installed already: MSVC++ 2015-2019 Redistributable (x86) - 14.20.27508
4. Rest of the installation runs smooth
5. Starting up QMS, checking actual (Default) TimeZone Settings: it is set to UTC
6. adding with drag&drop gpx file test-crash.gpx works perfectly
7. deleting track from workspace again
8. changing TimeZone Settings in QMS to 'automatic'
9. adding with drag&drop gpx file test-crash.gpx gives the same problem as on my initial private Windows 10 installation -> Track is not added properly and QMS needs restart to be able to use it again

As this is a complete different Windows 10 machine with no similarities regarding other installed applications (not even setup by myself) I am quite sure it can't be a 'local' problem only on my machine.
I can't explain myself why I have the problem on two different Windows 10 boxes but @wthaem can't reproduce it. He's running actually in the same timezone as I am locally. The only difference might be the region settings, but even if I change to Germany to have the comma instead of the dot as decimal separator, the problem still happens...

... I'm a bit puzzled

**Comment by wthaem** [comment]
Followed exactly your description - evrything runs as expected.

Possible differences of my installation:
* different MSVC - don't know which one is used on my machine, but it is a rather new one,
* QMS is not the release version, but the current development version compiled with Qt 5.15.0, GDAL 2.4.4, Proj 6.3.2,
* no map of Argentina, map window is empty.

**Comment by keenonkites** [comment]
@wthaem 

Regarding MSVC:

- I've upgraded from the above mentioned version MSVC++ 2015-2019 Redistributable (x86) - 14.20.27508 to the newest version 14.26.28720.3 for both x86 and x64 -> no success, same problem
- I've deinstalled this version and installed the version that comes in the QMS Installer (only for 2017 and only 64bit) -> even worse, QMS was not able to start on this windows 10
- Windows 10 is now also updated to the latest patch level

Regarding QMS Version:
- I'm running the normal one, downloaded the installer from github

Regarding Argentina map
- on the installation I'm trying out today there's no map at all installed

**Comment by kiozen** [comment]
For me it looks like some timestamp failure converting the UTC timestamps into the Argentinean timezone. However this is working on our systems. There might be something missing in the Windows installer but I wouldn't know what.

You could try to compile QMapShack on your own as a debug version. If the error persists you will be able to find the root cause. But chances are very high that this version behaves flawless.

**Comment by keenonkites** [comment]
Thanks, I've quickly read the compile instructions, quite intensive...

I've noticed the following sentence in Chapter I4:

> does not have any dependencies out of its own installation directory. So instead of creating an NSIS installer in the next step you could just copy this directory to an other machine for a quick deployment"

My first though was: if one of you guys (@wthaem most likely, as I'd need the windows version) can copy me your 'installation directory' and provide access to it (GoogleDrive ? Or any other download), then I could check a working installation on an actually non working machine. This would be another hint into direction of 'local machine problem' or 'Windows installer Problem' or anything else.

And it would possibly also give me a working version on my windows (where I have all gxp related data located) for Argentinian tracks without having to rethink/recalculate the tracktimes back into Argentinian time in my head... ;-)

**Comment by kiozen** [comment]
Yes it's quite intensive as it is really step by step with every detail. But that makes it easy to follow without background knowledge.

We could try this. However if it fails you wouldn't be able to debug.

**Comment by wthaem** [comment]
There is now one more PC, where QMS crashes with the Argentina file!

What I did:

* I took a PC that never has seen QMS and where not much software is installed.
* I installed QMS using the installation package with default settings, i.e. MSVC was installed too.
* Restart PC.
* Start of QMS, add some map to have proper map window. No change of timezone in QMS!
* Load `test-crash.gpx` - everything looks nice.
* Close QMS and restart: the QMS GUI can be seen for a short moment, then crash.
* Debugging is not possible on this machine! 

Last debug file entries:

~~~
2020-07-29 13:35:04.298 [debug] CGisItemTrk::updateVisuals() "CrashTest" "setupInterpolation()"
2020-07-29 13:35:04.299 [debug] CGisItemTrk::updateVisuals() "CrashTest" "deriveSecondaryData()"
~~~
 
I attach the INI file created by QMS in the Windows registry where the QMS settings can be found.

[qms.zip](https://github.com/Maproom/qmapshack/files/4994646/qms.zip)

**Comment by keenonkites** [comment]
@wthaem 
... that makes me feeling a little bit better: seems to be a very specific hidden problem inside the 'normal installation' and not something specific to my machines. 

I've checked the QMS Windows registry file you sent:
`[HKEY_CURRENT_USER\SOFTWARE\QLandkarte\QMapShack\Units\timezone]
"mode"=dword:00000002`
This means 'Automatic' timezone, which fits the the findings so far.

Would you mind zipping and providing me your self-compiled QMS for testing a running version directly on a 'problematic' machine ? Even if it will not allow me to debug deeper we can still crosscheck/confirm if the problem really is inside the installation package.... or (only thinking loud now) inside some missing setting inside the registry or similar.

**Comment by wthaem** [comment]
> Would you mind zipping and providing me your self-compiled QMS for testing a running version directly on a 'problematic' machine ? 
No. But my limited upload speed would be a serious obstacle for sending about 300 MB. And after all: its software running on a Windows system. I never know the ingredients the system adds to my software from some system folder.

I'll try to reinstall the original release package on my working machine. If the crash happens on this machine too, then I can try to debug the release version.

**Comment by keenonkites** [comment]
Sounds like a deal... If I can be of any help, just let me know.

**Comment by wthaem** [comment]
I got the crash on my working machine with the release package but only in the second run! Starting QMS with an existing workspace.db didn't result in a crash. Removing the old workspace.db and restarting as described reproduced the crash.

@kiozen: What is now the best way to proceed? I'm not sure, if I can re-activate all the packages used (Qt, proj, ...) on the level of the release. BTW: Qt5Core.dll is the DLL causing the crash.

**Comment by kiozen** [comment]
any chance to get a backtrace? The crash might be in Qt5Core.dll but the root cause has to be in the code of QMapShack. If I would know what line is causing the crash I would have a chance to fix it.

**Comment by wthaem** [comment]
Can this be done with the release version without compiling a debug version? 

Has recompilation of the release version to start with a checkout of the branch master?

The release was compiled with Qt 5.12.8 which isn't installed on my computer. Qt doesn't offer this version anymore for download. Which Qt version should be used instead of 5.12.8? A newer one? An older one?

**Comment by kiozen** [comment]
No, the release version will not do as it misses symbol information.

Another option you would have is to add debug output. From the logs the problem might be in `deriveSecondaryData()`. And obviously it has to do something with the timestamps. Therefore adding debug output to everything that concerns timestamps outputting the used variables might help.

Can't tell about the version. I would start with an older one.

**Comment by kiozen** [comment]
Strange indeed. Maybe it's really just Qt 5.12.8 or something went wrong when packaging the installer.

**Comment by huirad** [comment]
@kiozen made me aware of this issue.
I tried with my QMS 1.15.0 installation, added some tracks from Peru.
All good. 
Then I changed the time zone to Buenos Aires - now it's crashing.
Even if I start QMS again, it's always crashing. Baaaad!

**Comment by keenonkites** [comment]
Just tested it with new Release V1.15.1 on same Windows as before: same behaviour. Hanging again while adding test gpx when using 'automatic' timezone (Or by choosing BuenosAires Timezone directly).

Thought it's worth a try...
